making verdi setup & verdi quicksetup more user friendly

ltalirz opened the pull request:
Please see the commit message for the small improvement to verdi quicksetup.

A few more things that are still to do:
* Currently, `verdi setup` provides a default email address (`aiida@localhost`), while `verdi quicksetup` doesn't. If at all, it should be the other way around, but I guess a default email address is not a good idea in either case as it will mean that everybody uses the default

* Currently, `verdi setup` does not ask for first name, last name and institution, while `verdi quicksetup` does. Again, I think `verdi setup` should ask for these as well(?)

ltalirz commented:
Ok, I've now realised that `verdi setup` *does* ask for name and institution, but only if one modifies the email address from the default one (and only after the profile repository and the database have already been created).

So, would it make sense to remove this default value or is it needed for something?

Also, it seems the prompts for the user-properties in `cmdline/commands/user.py` don't use 'click' and have thus a different way of handling default values (click shows the default value in brackets in the prompt, while for the user email the default text is simply copied into the input field). Using a consistent way would make it clearer for the user.

giovannipizzi commented:
I think the default email can be removed everywhere.

I agree that the UI should be consistent, indeed @DropD is in the process of converting everything to click (it's in one of the repositories under his account on GitHub) but this requires some time. Anyone is very welcome to help in the conversion process!

ltalirz commented:
And one more question: 
Currently, `verdi help quicksetup` shows the python docstring of the quicksetup function.
I think the original docstring "Uses click for options..." illustrates that this can lead to dilemmas, since what the person typing `verdi help quicksetup` needs to see is not necessarily what a programmer of aiida should see.
In particular, the person typing `verdi help quicksetup` should not only get a description of what the command does but also a dynamically generated list of options.
For the moment, I have simply added instructions on how to get a list of these options to the docstring, but this is not ideal. Any ideas for a better solution?

ltalirz commented:
I've added an easy way to have the `click` options appear in the output of `verdi help ...`. 
Furthermore, I have now removed the default email address from `verdi setup`.
Everything alright so far?

There is a further quirk I ran into: Currently, `verdi setup` provides a non-interactive mode, where everything is provided via the command line (I wanted to use this to set up a `test` profile to automatically run the tests after installing aiida).
However, `verdi setup` isn't able to do this: it can only connect to an existing database.
`verdi quicksetup` *can* set up the database, *but* it doesn't provide a fully non-interactive mode: even, when you provide all the information on the command line, it still asks 

`The default profile for the 'verdi' process is set to 'quicksetup_ltal': do you want to set the newly created 'test_aiida' as the new default? (can be reverted later) [y/N]: The default profile for the 'daemon' process is set to 'quicksetup_ltal': do you want to set the newly created 'test_aiida' as the new default? (can be reverted later) [y/N]:`

So, in my opinion
 * `verdi quicksetup` should simply be a wrapper to `verdi setup`, which fills most of the required variables with default values. 
 * both commands should be able to set up the database + database user (for `verdi setup` one might need to enable a switch `--create-db` that is on by default in `verdi quicksetup`)

Do you agree?

giovannipizzi commented:
Hi Leo!
I approved the PR. Do you want to merge it already or fix the things of your last comment? In the first case, you can already do it (after updating your branch). Otherwise, you can continue committing here.

Regarding the comment:
- I would add another click option to decide whether the new profile should be the default one or not, so everything can be non-interactive (and have click ask for Y/N if not provided)
- I agree quicksetup should be a wrapper
- I agree both should be able to work fully non-interactively
- I like the idea of having --create-db as an option of setup (and I guess this means moving the DB-creation logic in there from quicksetup)
- I'm not sure quicksetup can be masde as simple as just giving default values to setup. Quicksetup has to make some choices when trying to create the DB, and might need to have to ask the user for info, passwords, ...

Maybe Quicksetup should just contain the logic for creating the DB, and then it would call the normal setup with the proper options (depending on what it created?)

ltalirz commented:
@DropD
I have had a look at your repo and it seems you have done some quite extensive reorganisation of the code. Could you please give us a summary of what you did and what the current status is?

I am happy to help with the integration of your work into the aiida develop branch.
We should avoid postponing this merge too far into the future, though - improvements to the verdi commands are needed and we should implement them soon.

ltalirz commented:
Ok, so as discussed I have added a click option for quicksetup to avoid the prompt for making the new profile default in shell/daemon.
The other things we discussed require more rearrangement in the code and have therefore been postponed until @DropD is back.

I have added one more thing that I have been missing, which is "verdi profile delete myprofile".
This is my first time adding a slightly more significant amount of new lines, so please check it and let me know if I should do some things differently.

ltalirz commented:
I just realised I forgot to delete the file repository.
I guess I should take the path from 'AIIDADB_REPOSITORY_URI' right?
What is the best way to delete this, as in the future the repository may be on object store...

Also perhaps some of the functions I am using here (in particular the functions to manipulate the database) should probably not be inside quicksetup but rather in aiida.common.setup (?)

ltalirz commented:
Finished a first implementation of verdi profile delete.
Example command line usage:

```
$ verdi profile list
* default
* test_mynewtest
* test_aiida
* my_test
> tdevelop (DEFAULT) (DAEMON PROFILE)
(aiida_ltal) leopold@tsf-428-wpa-7-059:~/.aiida

$ verdi profile delete test_aiida
Delete associated database 'test_aiida'?
WARNING: All data will be lost. [y/N]: y
Deleting database 'test_aiida'.
Delete database user 'test_aiida'?
WARNING: This user might be used by other profiles as well. [y/N]: y
Deleting user 'test_aiida'.
Delete associated file repository '/Users/leopold/.aiida/repository-test_aiida/'?
WARNING: All data will be lost. [y/N]: y
Deleting directory '/Users/leopold/.aiida/repository-test_aiida/'.
Delete configuration for profile 'test_aiida'?
WARNING: This permanently removes the profile from the list of AiiDA profiles. [y/N]: y
Deleting configuration for profile 'test_aiida'.
```

DropD commented on aiida/cmdline/commands/profile.py:
```diff
@@ -135,3 +136,71 @@ def profile_list(self, *args):
+            elif users.count(user) > 1:
+                print("Associated database user '{}' is used by other profiles "\
+                      "and will not be deleted.".format(user))
+            elif click.confirm("Delete database user '{}'?".format(user)):
```
Same consideration as above (--force) applies

DropD commented on aiida/cmdline/commands/profile.py:
```diff
@@ -135,3 +136,71 @@ def profile_list(self, *args):
+            elif not os.path.isdir(repo_path):
+                print("Associated file repository '{}' is not a directory."\
+                       .format(repo_path))
+            elif click.confirm("Delete associated file repository '{}'?\n" \
```
Another case for --force

DropD commented on aiida/cmdline/commands/profile.py:
```diff
@@ -135,3 +136,71 @@ def profile_list(self, *args):
+                               "WARNING: All data will be lost.".format(repo_path)):
+                print("Deleting directory '{}'.".format(repo_path))
+                import shutil
+                shutil.rmtree(repo_path)
```
Maybe it would make sense to already delegate the task of deleting to another function in a more obvious place which will do case handling for file system vs. object store / other future options - repositories.

ltalirz replied:
For the moment, I've created an issue. 
https://github.com/aiidateam/aiida_core/issues/695

Will use more general functionality once it becomes available.

DropD commented on aiida/cmdline/commands/profile.py:
```diff
@@ -135,3 +136,71 @@ def profile_list(self, *args):
+                import shutil
+                shutil.rmtree(repo_path)
+
+            if click.confirm("Delete configuration for profile '{}'?\n" \
```
Another --force case

DropD commented on aiida/cmdline/verdilib.py:
```diff
@@ -805,17 +822,24 @@ def _quicksetup_cmd(self, email, first_name, last_name, institution, backend, db
+            if _set_default:
                 set_default_profile(process, profile_name, force_rewrite=True)
 
+    #TODO (ltalirz) db-related functions should be moved to the base level of verdilib
```
Please add an issue for this (instead of a comment) and assign it to me unless you are interested in fixing it yourself

ltalirz replied:
Done.

ltalirz commented on aiida/cmdline/commands/profile.py:
```diff
@@ -135,3 +136,71 @@ def profile_list(self, *args):
+            try:
+                profile = profiles[profile_to_delete]
+            except KeyError:
+                raise ValueError('Profile "{}" does not exist'.format(profile_to_delete))
```
right, I'm going for the 3rd option

ltalirz commented on aiida/cmdline/commands/profile.py:
```diff
@@ -135,3 +136,71 @@ def profile_list(self, *args):
+            db = profile.get('AIIDADB_NAME', '')
+            if not q._db_exists(db, pg_execute, **dbinfo):
+                print("Associated database '{}' does not exist.".format(db))
+            elif click.confirm("Delete associated database '{}'?\n" \
```
Completely agree, I am going with '--yes' (in analogy with apt-get).
For the moment, I check by hand whether '--yes' is in the argument list as this is the way it is done with '--color' in another function in the same file. 
To be replaced by more robust code in the future.

ltalirz commented:
@DropD Thanks for the constructive suggestions, let me know what you think

szoupanos commented:
One question/comment on this.

Since we going to click for modularity and easy testing of the CLI, wouldn't it be good to start adding some tests? Especially for new commands that are developed with click?

This will make our lives easier in the future.
